### PR TITLE
feat!: move to ESM first with dual ESM/CJS subpackage exports

### PR DESCRIPTION
**Pull request by userquin**
## Related Issues or Discussions

Talk with Kato and Moha in discord.

## Summary

Move to ESM first and include subpackage exports.

This PR also includes:
- now using rollup-plugin-dts  instead typescript to generate dts files
- remove `useProxy` from `src/macro.ts` module and patch the d.ts and d.cts files
- change `build` to run `build:all`: now we can generate multiples outputs (d.mts, d.cts and d.ts) and also the ts3.4 version, we need to play with the target and maybe lib to fix ts3.4
- ts3.4 now inside dist folder: we should include only dist in files (package.json)
- `postbuild`now includes only `yarn copy`, I've left old `xpostbuild`, we should remove it later
- `copy` script now copy src and esm from root, I've left old `xcopy`, we should remove it later: why copying src folder to remove it in the next script?

## Check List

- [x] `yarn run prettier` for formatting code and docs

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,49 +1,67 @@
 {
   "name": "valtio",
+  "type": "module",
```
Comment (dai-shi) [inline package.json:3]: ugh, unfortunately, this is not acceptable. we try to keep cjs first. we plan esm first for valtio v3.

Comment (userquin) [reply package.json:3]: This pr is for main, I can send another pr using commonjs to v2 branch (this PR should be closed)

**Comment by dai-shi** [comment]
> I can send another pr using commonjs to v2 branch (this PR should be closed)

Sounds good. This PR introduces too many changes and can't even see if there are breaking changes or not. I'm pretty sure it breaks some old environments.

I would appreciate if you can send small PRs. For example: #882 (well, it's exceptionally small.)

File: package.json [hunk 2]
```diff
@@ -1,49 +1,67 @@
...
+      "macro": ["./dist/ts3.4/macro.d.ts"],
+      "macro/vite": ["./dist/ts3.4/macro/vite.d.ts"],
+      "react": ["./dist/ts3.4/react.d.ts"],
+      "react/utils": ["./dist/ts3.4/react/utils.d.ts"],
+      "utils": ["./dist/ts3.4/utils.d.ts"],
+      "vanilla": ["./dist/ts3.4/vanilla.d.ts"],
+      "vanilla/utils": ["./dist/ts3.4/vanilla/utils.d.ts"],
+      "*": ["./dist/ts3.4/*"]
     },
-    "./*": {
```
Comment (dai-shi) [inline package.json:34]: I'm curious why we need to avoid subpath patterns?

**Comment by dai-shi** [comment]
> small PRs

At least, we should separate for different issues: a) fix the result in `dist` and b) fix for tests, which may depend on a).

**Pull request closed without merging**